app/models/energy/ScheduleBase.py

```diff
@@ -0,0 +1,100 @@
+"""Schedule Base Schema"""
+from pydantic import BaseModel, Schema, validator, ValidationError, UrlStr, ConstrainedStr
+from typing import List, Union
+from enum import Enum
+from uuid import UUID, uuid4
+from datetime import datetime
+
+
+class ScheduleContinuous(BaseModel):
+    """This Numeric Type allows all numbers, including fractional amounts, within the range to
+  be valid."""
+
+    type: Enum('ScheduleContinuous', {'type': 'ScheduleContinuous'})
+
+    schedule_continuous: float = None
+
+
+class ScheduleDiscrete(BaseModel):
+    """This Numeric Type allows all only integers within the range to be valid."""
+
+    type: Enum('ScheduleDiscrete', {'type': 'ScheduleDiscrete'})
+
+    schedule_discrete: int = None
+
+class ScheduleNumericType (BaseModel):
+    """Designates how the range values are validated."""
+
+    numeric_type:  Union[ScheduleContinuous, ScheduleDiscrete]
+
+
+class ScheduleUnitType (str, Enum):
+    dimensionless = 'Dimensionless'
+    temperature = 'Temperature'
+    delta_temperature = 'DeltaTemperature'
+    precipitation_rate = 'PrecipitationRate'
+    angle = 'Angle'
+    convection_coefficient = 'ConvectionCoefficient'
+    activity_level = 'ActivityLevel'
+    velocity = 'Velocity'
+    capacity = 'Capacity'
+    power = 'Power'
+    availability = 'Availability'
+    percent = 'Percent'
+    control = 'Control'
+    mode = 'Mode'
+
+
+class Date(BaseModel):
+    """Date."""
+
+    month: int = Schema(
+        1,
+        description='A value for month between `1`-`12`. Default is `1`.',
+        ge=1,
+        le=12
+    )
+
+    day: int = Schema(
+        1,
+        description='A value for day between `1`-`31`. Default is `1`.',
+        ge=1,
+        le=31
+    )
+
+
+class Time(BaseModel):
+    """Time."""
+
+    hour: int = Schema(
+        0,
+        description='A value for hour between `0`-`24`. Default is `0`.',
+        ge=0,
+        le=24,
```

Comment:
We do not follow EnergyPlus's convention where time starts counting from 1 and goes to 24.  We did this in legacy and realized that it is a huge mistake.  Now we start counting from 0:00 and we go to 23:59.  So this should be from 0 to 23.

Reply:
@chriswmackey I agree that time starts at 0:00 each day (it is that way in E+ and OpenStudio too).  However, I do think that the way schedules are specified by the end time (value is until time) suggest that you would want to report the value goes until 24:00 (which is really 0:00 of the next day).  For example:

`value: 1 until: 24:00` would apply the value 1 all times of the day

`value: 1 until: 23:59` would apply the value until 11:59 PM and then have an undefined value for the last minute of the day

it really is an artifact of the way that the values are over an interval, and the interval is specified by the end time of the interval.

Reply:
Thanks for clarifying, @macumber. So, am I correct in understanding that E+ will throw and error if someone inputs `value until 24:01` and, if we wanted to support formatting similar to E+, we need to have an extra check to make sure that the user does not do this?

Reply:
Does this also mean that E+ expects the end time of a previous interval and the start time of the next interval to be the same? If so, what value does E+ actually use for this timestep that is shared by both intervals (the previous one or the next one)? This lack of clarity makes me think that we might not want to follow this convention just because it is implicit rather than explicit and it is not the convention that we are following across the rest of Ladybug + Honeybee. Maybe we should just handle the conversion from real time to E+ overalpped time in the measure. Or we need to have methods that make it very clear to the user what is goig on. @mostapha, what do you think here?

Reply:
Correct, value until 24:01 is an error, custom validation can say if hr = 24 then min must = 0

The E+ schedule convention is not that bad, E+ is simulating energy over discrete time intervals so it just needs to know when the interval begins and ends

![image](https://user-images.githubusercontent.com/1327850/60359297-eb65b980-9995-11e9-831d-0092ccf77207.png)

Reply:
Ok, I guess that we will have to support 24:00 as a valid time in order to deal with this quirk of EnergyPlus.  This also means that we will have to add a validator in the Time object of the schema that permits 24:00 but not anything after (see my comment above about using Python's builtin `datetime.time` class for this).

I just feel the need to state that the way EnergyPlus schedules work is not intuitive and is not the way that we handle timeseries in the rest of Ladybug.  Most users expect that when a schedule says that people occupy an office from 9:00 unit 17:00, they expect that people will be in the office at 9:00 but that they might not be there precisely at 17:00 because they are only supposed to be there until 17:00 (right at 17:00, they should leave).  This is completely against how EnergyPlus schedules and climate data work and I realize that much of this goes back to poor design decisions made about the EPW format over 20 years ago.  But, while non-intuitive design is bad, it is a greater sin to be unclear with the user about what is actually happening in the simulation.  So, for the schema, we will forego the idea of having a `time until` that is a minute less than the next timestep.  We will set the `time until` for the previous interval to be the same as `time from` in the next interval.  And we will work to be as clear with the user as we can that their schedule inputs are likely not what they expect.

This includes several docstrings noting that that the `time from` part of the interval is never actually considered a part of the interval by EnergyPlus (only the timestep after the `time from`). Likewise, `time until` docstrings should be clear that the interval is not actually until that timesetp but rather it is until the timestep following the `time until`.
On the honeybee_energy Python end, we will have to make methods that show exactly what the schedule means timestep-by-timestep so that  users can be made aware of the exact value used at each datetime.  Lastly, on the Python end, we will have to continue the practice that we currently use in Ladybug for timeseries EPW data in that we take the last timestep of the data at the fictional time of 24:00 and move it to be 0:00 for the first timestep of the data.  We will have to do this for all hourly and sub-hourly simulation results of EnergyPlus if we have any hope of correctly aligning it with other timeseries information like solar positions, the output of daylight simulations, or just our own clocks.